Charon/filetypes/OpenPackagingConvention.py

```diff
@@ -0,0 +1,574 @@
+                if metadata_file not in self.zipfile.namelist(): #The metadata file is unknown to us.
+                    continue
+
+                metadata = json.loads(self.zipfile.open(metadata_file).read().decode("utf-8"))
+                if metadata_file == self.global_metadata_file: #Store globals as if coming from root.
+                    metadata_file = ""
+                elif metadata_file.endswith(".json"): #Metadata files should be named <filename.ext>.json, meaning that they are metadata about <filename.ext>.
+                    metadata_file = metadata_file[:-len(".json")]
+                self._readMetadataElement(metadata, metadata_file)
+
+        if self.mode != OpenMode.WriteOnly and not self.getMetadata("/3D/model.gcode"):
+            try:
+                # Check if the G-code file actually exists in the package.
+                self.zipfile.getinfo("/3D/model.gcode")
+            except KeyError:
+                return
+
+            gcode_stream = self.zipfile.open("/3D/model.gcode")
+            header_data = GCodeFile.parseHeader(gcode_stream, prefix = "/3D/model.gcode/")
+            self.metadata.update(header_data)
+
+    ##  Reads a single node of metadata from a JSON document (recursively).
+    #   \param element The node in the JSON document to read.
+    #   \param current_path The path towards the current document.
+    def _readMetadataElement(self, element: Dict[str, Any], current_path: str) -> None:
+        for key, value in element.items():
+            if isinstance(value, dict):  # json structures stuff in dicts if it is a subtree.
+                self._readMetadataElement(value, current_path + "/" + key)
+            else:
+                self.metadata[current_path + "/" + key] = value
+
+    ##  At the end of writing a file, write the metadata to the archive.
+    #
+    #   This should be written at the end of writing an archive, when all
+    #   metadata is known.
+    #
+    #   ALWAYS WRITE METADATA BEFORE UPDATING RELS AND CONTENT TYPES.
+    def _writeMetadata(self) -> None:
+        assert self.zipfile is not None
+        
+        keys_left = set(self.metadata.keys()) #The keys that are not associated with a particular file (global metadata).
+        metadata_per_file = {} # type: Dict[str, Dict[str, Any]]
+        for file_name in self.zipfile.namelist():
+            metadata_per_file[file_name] = {}
+            for metadata_key in self.metadata:
+                if metadata_key.startswith(file_name + "/"):
+                    # Strip the prefix: "/a/b/c.stl/print_time" becomes just "print_time" about the file "/a/b/c.stl".
+                    metadata_per_file[file_name][metadata_key[len(file_name) + 1:]] = self.metadata[metadata_key]
+                    keys_left.remove(metadata_key)
+        # keys_left now contains only global metadata keys.
+
+        global_metadata = {key:self.metadata[key] for key in keys_left}
+        if len(global_metadata) > 0:
```

Comment:
You don't need to use len here, you can just use if not global_metadata (empty dicts evaluate to False)